plugin/main.py

```diff
@@ -0,0 +1,279 @@
+    buf.add("# source: {}", proto_file)
+    buf.add("# plugin: {}", __name__)
+    if not services:
+        return buf.content()
+
+    buf.add("import abc")
+    buf.add("import typing")
+    buf.add("")
+    buf.add("import {}", const.__name__)
+    buf.add("import {}", client.__name__)
+    buf.add("import {}", exceptions.__name__)
+    buf.add("if typing.TYPE_CHECKING:")
+    with buf.indent():
+        buf.add("import {}", server.__name__)
+
+    buf.add("")
+    for mod in imports:
+        buf.add("import {}", mod)
+    for service in services:
+        if package:
+            service_name = "{}.{}".format(package, service.name)
+        else:
+            service_name = service.name
+        buf.add("")
+        buf.add("")
+        buf.add("class {}Base(abc.ABC):", service.name)
+        with buf.indent():
+            for name, _, request_type, reply_type in service.methods:
+                buf.add("")
+                buf.add("@abc.abstractmethod")
+                buf.add(
+                    "async def {}(self, stream: '{}.{}[{}, {}]') -> None:",
+                    name,
+                    server.__name__,
+                    server.Stream.__name__,
+                    request_type,
+                    reply_type,
+                )
+                with buf.indent():
+                    buf.add("pass")
+            buf.add("")
+            buf.add("def __mapping__(self) -> typing.Dict[str, {}.{}]:", const.__name__, const.Handler.__name__)
+            with buf.indent():
+                buf.add("return {{")
+                with buf.indent():
+                    for method in service.methods:
+                        name, cardinality, request_type, reply_type = method
+                        full_name = "/{}/{}".format(service_name, name)
+                        buf.add("'{}': {}.{}(", full_name, const.__name__, const.Handler.__name__)
+                        with buf.indent():
+                            buf.add("self.{},", name)
+                            buf.add("{}.{}.{},", const.__name__, const.Cardinality.__name__, cardinality.name)
+                            buf.add("{},", request_type)
+                            buf.add("{},", reply_type)
+                        buf.add("),")
+                buf.add("}}")
+
+        buf.add("")
+        buf.add("")
+        buf.add("class Unimplemented{}Base({}Base):", service.name, service.name)
+        with buf.indent():
+            for name, _, request_type, reply_type in service.methods:
+                buf.add("")
+                buf.add(
+                    "async def {}(self, stream: '{}.{}[{}, {}]') -> None:",
+                    name,
+                    server.__name__,
+                    server.Stream.__name__,
+                    request_type,
+                    reply_type,
+                )
+                with buf.indent():
+                    buf.add("raise grpclib.exceptions.GRPCError(grpclib.const.Status.UNIMPLEMENTED)")
```

Comment:
These are the lines I added compared to the [grpclib](https://github.com/vmagamedov/grpclib/blob/master/grpclib/plugin/main.py) version